app/src/main/java/org/breezyweather/remoteviews/gadgetbridge/GadgetBridgeService.kt

```diff
@@ -52,17 +53,24 @@ object GadgetBridgeService {
     private const val GB_INTENT_PACKAGE_SMARTSPACER = "nodomain.pacjo.smartspacer.plugin.genericweather"
     private const val GB_INTENT_ACTION = "$GB_INTENT_PACKAGE.ACTION_GENERIC_WEATHER"
 
-    fun sendWeatherBroadcast(context: Context, location: Location) {
-        if (location.weather?.current == null) {
+    fun sendWeatherBroadcast(context: Context, locations: List<Location>) {
+        if (locations[0].weather?.current == null) {
```

Comment:
@papjul I just noticed that, even on the main branch, this will get called multiple times in succession when the user has multiple locations configured. In main, it will only contain the primary location, while in this branch it gets called first with 1 location, then 2.

I am not yet fully comfortable with the Breezy Weather codebase to understand whether this is something that we can easily prevent on the Breezy Weather side (only broadcast one time once we have updated data for all the locations), or whether we should try and deal with that on the Gadgetbridge side, although we'll have to implement some sort of delay before acting on the data, since we have no real way of knowing that the intent we received is the final one.

Do you have any thoughts on this?

Reply:
When you say "multiple times", do you mean twice or for as many times as there are locations?

In which case(s) does it happen?
- [ ] When setting is enabled
- [ ] When user manually refreshes any location
- [ ] When it's refreshed in the background

Reply:
Data below is with 3 locations configured (current + 2 fixed):

On main branch:
- Opening the app: 2x
- Pull down to refresh: 1x
- Enabling setting: 1x
- When it's refreshed in the background: need to confirm this

On this branch:
- Opening the app: 4x! One of them with only 1 location (order was 3, 3, 1, 3)
- Pull down to refresh: 1x with all 3 locations
- Enabling setting: 1x with just 1 location
- When it's refreshed in the background: also need to confirm 

I've also noticed that going back from settings to the main screen also triggers sending the data.

Overall, it does look like this branch makes it worse (possibly by the change done in the WeatherUpdateJob.kt).

Reply:
> I've also noticed that going back from settings to the main screen also triggers sending the data.

The main screen is made aware of any changes to locations made from everywhere in the app. When it receives the new list, it updates widgets, notifications, gadgets, etc. However, it is not aware that the updated location was made from an entrypoint that already take care of updating that.

It's a behavior inherited from Geometric Weather, which had way more unexpected updates in the background. I thought I had fixed all of them, but looks like there is some one remaining point, although I'm not sure how I can easily fix it.

> When it's refreshed in the background: need to confirm this

I wouldn't be surprised if you end up with the same behavior as above (updates a second time when you go back to main screen).

> Opening the app: 2x

I didn't expected that at all, could it be the same issue? Although I'm not sure why it would be more than once.

Reply:
I can confirm that the background refresh only triggers the broadcast once, with all 3 locations.

Regarding opening the app, it always triggers 2x, only sometimes 4 (but I can't reproduce this last one consistently).

I compared the stacktrace of both calls, but it's the same (async logic). Tracing it a bit more, it looks like `initView` in MainActivity is somehow called twice, but it's not clear to me why (maybe I'm going down the wrong rabbit hole).

```
org.breezyweather.remoteviews.gadgetbridge.GadgetBridgeService.sendWeatherBroadcast(GadgetBridgeService.kt:57)
org.breezyweather.remoteviews.Gadgets.updateGadgetIfNecessary(Gadgets.kt:27)
org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1.invokeSuspend$lambda$0(MainActivityViewModel.kt:615)
org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1.$r8$lambda$sWbMN4q7IH-eRq7FlwCxRLZPshM(Unknown Source:0)
org.breezyweather.main.MainActivityViewModel$refreshBackgroundViews$1$1$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
org.breezyweather.common.utils.helpers.AsyncHelper.delayRunOnIO$lambda$2(AsyncHelper.kt:56)
org.breezyweather.common.utils.helpers.AsyncHelper.$r8$lambda$5b3nePoL_LvCtLHM9YQ2MmyMu9Y(Unknown Source:0)
org.breezyweather.common.utils.helpers.AsyncHelper$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
io.reactivex.rxjava3.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onComplete(ObservableDoOnEach.java:133)
io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:288)
io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:194)
io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:256)
io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
java.lang.Thread.run(Thread.java:1012)
```

> The main screen is made aware of any changes to locations made from everywhere in the app.

Do keep in mind that I am not changing anything - just opening the main settings page and going back.

Reply:
Ah, that could be a regression from Geometric Weather then!
`lifecycleScope.launch` is definitely something we added to move to more Kotlin-style, but we still have some kind of mix between the two styles, as the "refresh when receiving data from background" is a code I don't understand, or at least that I don't know how to migrate. Additionally, I'm really not familiar with lifecycles so it doesn't help.

Receiving the data twice shouldn't be a problem (just not very efficient) if the data is the same. However, I seemed to understand that you received different data? Sometimes only 1 location?

Reply:
Yes, sometimes there is just 1 location. And yeah, the main concern would be efficiency - although I think we might be able to easily ignore duplicated broadcasts on the Gadgetbridge if the data is actually the same when all locations are sent (I did not double-check that it's exactly the same).

Reply:
There is indeed an issue with the `refreshBackgroundViews` function from the main activity, which receives a list of one location in some cases. I will look into fixing this in priority #871

Reply:
I actually I read the code a bit too fast, the `first` parameter didn't mean we take only the first location, it was the `first` from a `Pair`.
Can you provide in which case you get only one location?

Reply:
Starting the app with multiple locations always sends twice - the first with 1 location, the second with all:

Logs:

```
03-31 15:53:30.627 24857 24927 D BreezyWeather: Sending GadgetBridge data for 1 locations
03-31 15:53:30.761 24857 25602 D BreezyWeather: Sending GadgetBridge data for 3 locations
```

Disabling and re-enabling the setting to broadcast weather also consistently sends only 1 location.

Reply:
> Disabling and re-enabling the setting to broadcast weather also consistently sends only 1 location.

I commented on `RefreshHelper` file for that issue.

Let me know if it still happens on app opening after the fix.

Reply:
This fixes sending all locations when disabling and re-enabling the setting.

However, opening the app still sends 1 and then 3.

Reply:
This seems to be solved, as per https://github.com/breezy-weather/breezy-weather/pull/501#discussion_r1545732119